packages/cli-kit/src/plugins.ts

```diff
@@ -25,12 +27,46 @@ export async function lookupTunnelPlugin(plugins: Plugin[]): Promise<TunnelPlugi
  *
  * Responses are organised into a dictionary, keyed by plug-in name. Only plug-ins that have hooks registered for the given event, and the hooks were run successfully, are included.
  */
-export async function fanoutHooks<T extends keyof Hooks, TResult = Hooks[T]['return']>(
-  config: Config,
-  event: T,
-  options: Hooks[T]['options'],
+export async function fanoutHooks<TPluginMap extends HookReturnsPerPlugin, TEvent extends string & keyof TPluginMap>(
+  config: Interfaces.Config,
+  event: TEvent,
+  options: TPluginMap[typeof event]['options'],
   timeout?: number,
-): Promise<{[pluginName: string]: TResult}> {
+): Promise<Partial<TPluginMap[typeof event]['pluginReturns']>> {
   const res = await config.runHook(event, options, timeout)
-  return Object.fromEntries(res.successes.map(({result, plugin}) => [plugin.name, result]))
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return Object.fromEntries(res.successes.map(({result, plugin}) => [plugin.name, result])) as any
 }
+
+interface HookReturnsPerPlugin {
+  public_command_metadata: {
+    options: {[key: string]: never}
+    pluginReturns: {
+      '@shopify/app': Partial<
+        Pick<Schemas['app_cli3_command/1.0']['public'], 'project_type' | 'api_key' | 'partner_id'> & {
+          [key: string]: unknown
+        }
+      >
+      [pluginName: string]: {[key: string]: unknown}
```

Comment:
Only the things where we want to type-check what we're sending to monorail -- the highlighted line pulls the type definitions from the monorail schema, then we have `FanoutHookFunction` which can ensure that the hook handler in some particular plug-in includes these keys with the right type (they are *all* optional, always).

I think this will probably get iterated on a bit as more keys are added and start to understand where the annoying bits are!